File: WealthBuilder/frontend/src/utils/slug.ts
// Turns an asset name into a URL slug: lowercased, with every run of non-alphanumeric
// characters collapsed to a single dash and leading/trailing dashes trimmed. "Precious Metals"
// becomes "precious-metals". Slugifying both the asset name and the route segment makes the
// match inherently case- and spacing-insensitive.

export const slugify = (value: string): string => {
    return value
        .toLowerCase()
        .replace(/[^a-z0-9]+/g, '-')
        .replace(/^-+|-+$/g, '');
};

File: WealthBuilder/backend/src/main/java/com/wealthbuilder/backend/utils/Slug.java
package com.wealthbuilder.backend.utils;

import java.util.Locale;


/**
 * Turns an asset name into a URL slug, mirroring the frontend's {@code slugify} exactly so the
 * two sides agree: lowercased, every run of non-alphanumeric characters collapsed to a single
 * dash, leading/trailing dashes trimmed. "Precious Metals" becomes "precious-metals".
 */
public final class Slug {

    private Slug() {
    }

    public static String of(String value) {
        return value
                .toLowerCase(Locale.ROOT)
                .replaceAll("[^a-z0-9]+", "-")
                .replaceAll("^-+|-+$", "");
    }
}
